subsys/zbus/zbus.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+#define _ZBUS_ASSERT_UNREACHABLE __ASSERT(0, "Reached unreachable code")
+
+#if defined(CONFIG_THREAD_NAME)
+#define _ZBUS_CURRENT_THREAD_INFO k_thread_name_get(k_current_get()), k_current_get()
+#else
+#define _ZBUS_CURRENT_THREAD_INFO "_", k_current_get()
+#endif /* CONFIG_THREAD_NAME */
+
+int zbus_obs_set_enable(struct zbus_observer *obs, bool enabled)
+{
+	if (obs == NULL) {
+		LOG_ERR("Invalid arg, obs is NULL.");
+		_ZBUS_ASSERT_UNREACHABLE;
+		return -EINVAL;
+	}
+	obs->enabled = enabled;
+	return 0;
+}
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
+{
+}
+
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan)
+{
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->callback != NULL)) {
+			(*obs)->callback(chan);
+		}
+	}
+	_zbus_notify_runtime_listeners(chan);
+}
+
+int _zbus_notify_subscribers(struct zbus_channel *chan)
+{
+	int err = 0;
+
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put((*obs)->queue, &chan,
+					 K_MSEC(CONFIG_ZBUS_EVENT_DISPATCHER_QUEUE_TIMEOUT));
+			if (err) {
+				return err;
+			}
```

Comment:
A single timeout option for all channels forces users to always take the worst case scenario into account. Maybe turn this into a channel member or pass it as a parameter when publishing.

Also not sure about the early return, if a single subscriber wasn't ready to receive the message, shouldn't all the others at least get it?